### PR TITLE
Fixes InMemoryProfileService.update

### PR DESCRIPTION
**Pull request by eroux**
The `InMemoryProfileService.update()` function used to erroneously assume that all attributes were present during an update so it just replaced the previous profile entry with the asked one, overwriting attritutes that were not meant to be overwritten. This PR fixes this behavior and adds a test. I'll send a separate PR for CouchProfileService which is doing the same mistake.

### CODE REVIEW

**Comment by eroux** [comment]
@leleuj @victornoel any comment on this one and #959 ?

**Comment by leleuj** [comment]
I'm not convinced there is a mistake here: it's more a matter of what behavior we want for the update.

Previously, update meant erase if the profile already existed and create if it didn't.
What you propose is to merge the new profile and the existing one if it already exists and not to create it if it doesn't exist.

It's a breaking change also which fits badly with a minor version like v2.1.

What use case pushed you into that direction?

**Comment by eroux** [comment]
About the change: if you look at the [DbProfileService](https://github.com/pac4j/pac4j/blob/master/pac4j-sql/src/main/java/org/pac4j/sql/profile/service/DbProfileService.java#L103) it clearly just updates the asked fields without removing the others when in legacy mode.

About the use case: I just want to update a profile without updating the password. Or is there a way to do that that I missed?

**Comment by eroux** [comment]
BTW, the use case is reflected in the additions in the tests, something like

```java
// update with no password update
 inMemoryProfileService.update(profile, null);
 // check credentials
inMemoryProfileService.validate(credentials2, null);
CommonProfile profile4 = credentials.getUserProfile();
assertNotNull(profile4);
```

was failing before (with the correct credentials), but doesn't fail now.

**Comment by leleuj** [comment]
I guess you're right about the password. It would be erased if not provided during the update.

But if the profile does not already exist, the update won't be done, right?

**Comment by eroux** [comment]
No it would not, should it? Maybe adding some documentation on the function would help... I remember trying to find a path in the code where a profile would be updated without being created first, but I coudln't find any, so I assumed that update was always called on a profile already existing in the database... should it raise an exception if it isn't? or fail silently? or create the profile?

**Comment by leleuj** [comment]
My feeling is that we should create the profile at the update if it doesn't already exist to be backward compatible. But please add a TODO to keep track that we need to revisit that for the version 3.0.

**Comment by eroux** [comment]
Ok, I'll change that, but if the update behavior should be revisited, I guess it should be revisited for all plugins, so maybe opening a separate issue would be better than writing a TODO in this file?

**Comment by leleuj** [comment]
Yes, please open an issue. Forget about the TODO idea...

**Comment by eroux** [comment]
I've updated the PR with the requested change, is there anything else?

File: pac4j-core/src/main/java/org/pac4j/core/profile/service/InMemoryProfileService.java [hunk 1]
```diff
@@ -51,7 +51,12 @@ protected void insert(final Map<String, Object> attributes) {
     protected void update(final Map<String, Object> attributes) {
         final String id = (String) attributes.get(getIdAttribute());
         logger.debug("Updating id: {} with attributes: {}", id, attributes);
-        profiles.put(id, attributes);
+        final Map<String,Object> profile = profiles.get(id);
+        if (profile != null) {
+            profile.putAll(attributes);
+        } else {
+            insert(attributes);
```
Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/profile/service/InMemoryProfileService.java]: Cannot we just do: `profiles.put(id, attributes);`?

Comment (eroux) [inline pac4j-core/src/main/java/org/pac4j/core/profile/service/InMemoryProfileService.java]: if you prefer

**Comment by eroux** [comment]
done

**Comment by leleuj** [comment]
Thanks

**Comment by leleuj** [comment]
Can you do the same change for the other PR for CouchDB?

**Comment by eroux** [comment]
sure, I'll do that later today

**Pull request merged**